# Introduction to Python
- Python is a general purpose high level programming language.  
- Python was developed by Guido Van Rossam in 1989 while working at National Research Institute at Netherlands.
- It has simple easy-to-use syntax, making it the perfect language for someone trying to learn computer programming for the first time.
- In contrast to other popular languages such as C, C++, Java, and C#, Python strives to provide a simple but powerful syntax.
-  Python was made available to public in 1991. The official Date of Birth for Python is : Feb 20th 1991.  
- Python has wide range of applications from Web development (like: Django and Bottle), scientific and mathematical computing (Orange, SymPy, NumPy) to desktop graphical user Interfaces (Pygame, Panda3D).
- Python is used for software development at companies and organizations such as Google, Yahoo, Facebook, CERN, Industrial Light and Magic, and NASA. Experienced programmers can accomplish great
things with Python, but Python’s beauty is that it is accessible to beginning programmers and allows them
to tackle interesting problems more quickly than many other, more complex languages that have a steeper
learning curve.  
  
   
# Facts about Python
- The name Python was selected from the TV Show **"The Complete Monty Python's Circus"**, which was broadcasted in BBC from 1969 to 1974.
- Sir Guido developed Python language by taking almost all programming features from different languages:
    - Functional Programming Features from C
    - Object Oriented Programming Features from C++
    - Scripting Language Features from Perl and Shell Script
    - Modular Programming Features from Modula-3
    - Most of syntax in Python Derived from C and ABC languages.  
    
**General characteristics of Python:**

* **clean and simple language:** Easy-to-read and intuitive code, easy-to-learn minimalistic syntax, maintainability scales well with size of projects.
* **expressive language:** Fewer lines of code, fewer bugs, easier to maintain.  
  
**Technical details:**
  
* **dynamically typed:** No need to define the type of variables, function arguments or return types.
* **automatic memory management:** No need to explicitly allocate and deallocate memory for variables and data arrays. No memory leak bugs. 
* **interpreted:** No need to compile the code. The Python interpreter reads and executes the python code directly.  
  
# Uses of Python, Advantages and disadvantages  
1. For developing Desktop Applications
2. For developing web Applications
3. For developing database Applications
4. For Network Programming
5. For developing games
6. For Data Analysis Applications
7. For Machine Learning
8. For developing Artificial Intelligence Applications
9. For IOT  
  
**Advantages**  
* The main advantage is ease of programming, minimizing the time required to develop, debug and maintain the code.
* Well designed language that encourage many good programming practices:
 * Modular and object-oriented programming, good system for packaging and re-use of code. This often results in more transparent, maintainable and bug-free code.
 * Documentation tightly integrated with the code.
* A large standard library, and a large collection of add-on packages.  
  
**Disadvantages**  
* Since Python is an interpreted and dynamically typed programming language, the execution of python code can be slow compared to compiled statically typed programming languages, such as C and Fortran.   
  
# Features of Python
- **Simple and easy to learn**:
    - Python is a simple programming language. When we read Python program,we can feel like
reading english statements.
    - The syntaxes are very simple and only 30+ kerywords are available.
    - When compared with other languages, we can write programs with very less number of
lines. Hence more readability and simplicity.
    - We can reduce development and cost of the project.  
    
- **Open Source**:  
    - We can use Python software without any licence and it is freeware.
    - Its source code is open,so that we can we can customize based on our requirement.
    - **Eg:** Jython is customized version of Python to work with Java Applications.   
    
- **High Level Programming language**:
    - Python is high level programming language and hence it is programmer friendly language.
    - Being a programmer we are not required to concentrate low level activities like memory management and security etc..
- **Platform Independent**:
    - Once we write a Python program,it can run on any platform without rewriting once again.
- **Portability**:
    - Python programs are portable. ie we can migrate from one platform to another platform very easily. Python programs will provide same results on any paltform.
- **Dynamically Typed**:
    - In Python we are not required to declare type for variables. Whenever we are assigning the value, based on value, type will be allocated automatically.Hence Python is considered as dynamically typed language.
    - But Java, C etc are Statically Typed Languages because we have to provide type at the beginning only.
    - This dynamic typing nature will provide more flexibility to the programmer.
- **Both Procedure Oriented and Object Oriented**:
    - Python language supports both Procedure oriented (like C, pascal etc) and object oriented (like C++,Java) features. Hence we can get benefits of both like security and reusability etc.
- **Extensible**:
    - We can use other language programs in Python.
    - The main advantages of this approach are:
        - We can use already existing legacy non-Python code
        - We can improve performance of the application  
- **Embedded**:
    - We can use Python programs in any other language programs. i.e we can embedd Python programs anywhere.

# Keywords
- In Python some words are reserved to represent some meaning or functionality. Such type
of words are called Reserved words/Keywords.  
- There are 35 reserved words available in Python 3.7.x as of now  
  
```python
>>> import keyword
>>> keyword.kwlist
```
Note:
1. All Reserved words in Python contain only alphabet symbols.
2. Except the following 3 reserved words, all contain only lower case alphabet symbols.

In [3]:
import keyword
print(len(keyword.kwlist))
print(keyword.kwlist)

35
['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


# Identifiers
A name in Python program is called identifier.  
It can be class name or function name or module name or variable name.  
  
Rules to define identifiers in Python:  
    1. The only allowed characters in Python are
        - alphabet symbols(either lower case or upper case)
        - digits(0 to 9)
        - underscore symbol(_)  By mistake if we are using any other symbol like $ then we will get syntax error.
        - cash = 10  (CORRECT)
        - ca$h =20    (WRONG)
    2. Identifier should not starts with digit
        - 123total    (WRONG)
        - total123   (CORRECT)

    3. Identifiers are case sensitive. Of course Python language is case sensitive language.
```python
    >>> total=10
    >>> TOTAL=999
    >>> print(total) #10
    >>> print(TOTAL) #999
```  
# Quizz
Q. Which of the following are valid Python identifiers?   

- 123total (Wrong)  
- total123 (Correct)  
- java2share (Correct)  
- ca$h (Wrong)  
- _abc_abc_ (Correct)  
- def (Wrong)  
- if (Wrong)  
  
# Data Types
- Data Type represent the type of data present inside a variable.
- In Python we are not required to specify the type explicitly. Based on value provided,the type will be assigned automatically.Hence Python is Dynamically Typed Language.
- Python contains the following inbuilt data types:  
    
    - int
    - float
    - complex
    - bool
    - str
    - bytes
    - bytearray
    - range
    - list
    - tuple
    - set
    - frozenset
    - dict
    - None 
    
**Python contains several inbuilt functions**  
  
- **type()**
    - to check the type of variable  
- **id()**
    - to get address of object
- **print()**
    - to print the value  
    
**Python Memory Management**  
> In python Everything is an **object**   
  

![](img\mm.png)  
**int data type:**  
We can use int data type to represent whole numbers (integral values)  
  
In Python 3, there is effectively no limit to how long an integer value can be. Of course, it is constrained by the amount of memory your system has, as are all things, but beyond that an integer can be as long as you need it to be:  
  

In [1]:
a = 11111111111111111111111111111111111111111111111111111111111111111
print(a + 1)
print(type(a))

11111111111111111111111111111111111111111111111111111111111111112
<class 'int'>


Python interprets a sequence of digits without any prefix to be a decimal number:  
  
|     Prefix     | Interpretation | Base |    Allowed Digits  |   
|----------------|----------------|:----:|--------------------|  
| 0b   or  0B    | Binary         | 2    |     0 & 1          |  
| 0o  or  0O     | Octal          | 8    |     (0-7)          |  
| 0x  or  0X     | HexaDecimal    | 16   | (0-9) & (a-f, A-F) |    
  
The datatype type of a Python integer, irrespective of the base used to specify it, is called int:  
```python
>>> type(10)
<class 'int'>
>>> type(0o10)
<class 'int'>
>>> type(0x10)
<class 'int'>
```

In [4]:
a = 0B111
b = 0b010110
print(a)
print(b)
print(bin(a))
print(bin(b))

7
22
0b111
0b10110


In [25]:
a = 0o123
b = 0O746
print(a)
print(b)
print(oct(a))
print(oct(b))

83
486
0o123
0o746


In [28]:
a = 0XFACE
b = 0XBeef
c = 0XBee
print(a)
print(b)
print(c)
print(hex(a))
print(hex(b))
print(hex(c))

64206
48879
3054
0xface
0xbeef
0xbee


**float data type:**  
- We can use float data type to represent floating point values (decimal values)   
- Eg f = 1.234
- We can also represent floating point values by using exponential form (scientific notation)  
- Eg f = 1.2e3 (instead of 'e' we can use 'E')
- The main advantage of exponential form is we can represent big values in less memory.

In [6]:
a = 1.234
print(a)
print(type(a))
b = 1.2E3 # 1.2 * 10^3
print(b)
print(type(b))
c = 1.2e-4 # 1.2 * 10^(-4)
print(c)

1.234
<class 'float'>
1200.0
<class 'float'>
0.00012


Almost all platforms represent Python float values as 64-bit “double-precision” values, according to the [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754-2008_revision) standard. In that case, the maximum value a floating-point number can have is approximately 1.8 ⨉ 10<sup>308</sup>. Python will indicate a number greater than that by the string **inf**:

In [13]:
p = 1.79E308
print(p)
f = 1.8e308
print(f)

1.79e+308
inf


Floating point numbers are represented internally as binary (base-2) fractions. Most decimal fractions cannot be represented exactly as binary fractions, so in most cases the internal representation of a floating-point number is an approximation of the actual value. In practice, the difference between the actual value and the represented value is very small and should not usually cause significant problems.  
  
**Why can't decimal numbers be represented exactly in binary?**   
Floating-point numbers are represented in computer hardware as base 2 (binary) fractions. For example, the decimal fraction  
```python
0.125
```  
has value 1/10 + 2/100 + 5/1000, and in the same way the binary fraction   
```
0.001
```  
has value 0/2 + 0/4 + 1/8. These two fractions have identical values, the only real difference being that the first is written in base 10 fractional notation, and the second in base 2.  
Unfortunately, most decimal fractions cannot be represented exactly as binary fractions. A consequence is that, in general, the decimal floating-point numbers you enter are only approximated by the binary floating-point numbers actually stored in the machine.  
  
The problem is easier to understand at first in base 10. Consider the fraction 1/3. You can approximate that as a base 10 fraction:  
```python
0.3 or 0.33 or 0.333....
```  
and so on. No matter how many digits you’re willing to write down, the result will never be exactly 1/3, but will be an increasingly better approximation of 1/3.

In the same way, no matter how many base 2 digits you’re willing to use, the decimal value 0.1 cannot be represented exactly as a base 2 fraction. In base 2, 1/10 is the infinitely repeating fraction  
```python
0.0001100110011001100110011001100110011001100110011...
```  
Stop at any finite number of bits, and you get an approximation. On most machines today, floats are approximated using a binary fraction with the numerator using the first 53 bits starting with the most significant bit and with the denominator as a power of two.  
  
Many users are not aware of the approximation because of the way values are displayed. Python only prints a decimal approximation to the true decimal value of the binary approximation stored by the machine. On most machines, if Python were to print the true decimal value of the binary approximation stored for 0.1, it would have to display  
```python
>>> 0.1
0.1000000000000000055511151231257827021181583404541015625
```  
That is more digits than most people find useful, so Python keeps the number of digits manageable by displaying a rounded value instead  
  
To prove 0.1 is not exactly 1/10, summing three values of 0.1 may not yield exactly 0.3, either:  

In [45]:
print(format(0.1, '.40f'))
print(0.1)
print(0.1 + 0.1 + 0.1 == 0.3)

0.1000000000000000055511151231257827021182
0.1
False


Python provides tools that may help on those rare occasions when you really do want to know the exact value of a float. The float.as_integer_ratio() method expresses the value of a float as a fraction:

In [32]:
x = 3.14159
print(x.as_integer_ratio())

(3537115888337719, 1125899906842624)


Since the ratio is exact, it can be used to losslessly recreate the original value:

In [34]:
x == 3537115888337719 / 1125899906842624

True

In [35]:
y = 0.1
print(y.as_integer_ratio())

(3602879701896397, 36028797018963968)


In [36]:
y == 3602879701896397 / 36028797018963968

True

In [37]:
3602879701896397 / 36028797018963968

0.1

In [43]:
from math import fsum
print(0.1 == 0.1)

True


In [46]:
from decimal import Decimal
from fractions import Fraction

print(Fraction.from_float(0.1))
Fraction(3602879701896397, 36028797018963968)

print((0.1).as_integer_ratio())

print(Decimal.from_float(0.1))

3602879701896397/36028797018963968
(3602879701896397, 36028797018963968)
0.1000000000000000055511151231257827021181583404541015625


**Complex Number :**  
- A complex number is of the form **a + jb**  
- In the real part if we use int value then we can specify that either by decimal,octal,binary or hexa decimal form.  
- But imaginary part should be specified only by using decimal form.
- we can perform operations on complex type values.

In [33]:
a = 10 + 1.5j
b = 20 + 2.5j
c = a + b
print(c)
print(type(c))

(30+4j)
<class 'complex'>


Complex data type has some inbuilt attributes to retrieve the real part and
imaginary part


In [34]:
c = 10.5 + 3.6j
print(c.real)
print(c.imag)

10.5
3.6


**bool data type:**  
- We can use this data type to represent boolean values.
- The only allowed values for this data type are:
- True and False
- Internally Python represents True as 1 and False as 0   
  
```python
>>> type(True)
<class 'bool'>
>>> type(False)
<class 'bool'>
```

In [35]:
a=10
b=20
c = a < b
print(c)
print(True+True)
print(True-False)

True
2
1


**str type:**  
- str represents String data type.
- A String is a sequence of characters enclosed within single quotes or double quotes.   
  
```python
>>> print("I am a string.")
I am a string.
>>> type("I am a string.")
<class 'str'>
>>> print('I am too.')
I am too.
>>> type('I am too.')
<class 'str'>
```   
  
- By using single quotes or double quotes we cannot represent multi line string literals  
- For this requirement we should go for triple single quotes(''') or triple double quotes(""")   
- A string in Python can contain as many characters as you wish. The only limit is your machine’s memory resources. A string can also be empty:  
  
**What if you want to include a quote character as part of the string itself?**  
- If you want to include either type of quote character within the string, the simplest way is to delimit the string with the other type. If a string is to contain a single quote, delimit it with double quotes and vice versa:

In [48]:
print("This string contains a single quote 'Hi' character.")
print('This string contains a double quote "Hi" character.')

This string contains a single quote 'Hi' character.
This string contains a double quote "Hi" character.


#### Suppressing Special Character Meaning
```python
>>> print('This string contains a single quote (') character.')
SyntaxError: invalid syntax
```  
Specifying a backslash in front of the quote character in a string “escapes” it and causes Python to suppress its usual special meaning. It is then interpreted simply as a literal single quote character:  
  
```python
>>> print('This string contains a single quote (\') character.')
This string contains a single quote (') character.
```  
  
To break up a string over more than one line, include a backslash before each newline, and the newlines will be ignored:


In [53]:
print("a\
b\
c")

abc


To include a literal backslash in a string, escape it with a backslash:

In [54]:
print('foo\\bar')

foo\bar


#### Escape Sequence
<pre>
\a	        -   ASCII Bell (BEL) character  
\b	        -   ASCII Backspace (BS) character  
\f	        -   ASCII Formfeed (FF) character  
\n	        -   ASCII Linefeed (LF) character  
\N{name}	-   Character from Unicode database with given (name)  
\r	        -   ASCII Carriage Return (CR) character  
\t	        -   ASCII Horizontal Tab (TAB) character  
\uxxxx	    -   Unicode character with 16-bit hex value xxxx  
\Uxxxxxxxx  -   Unicode character with 32-bit hex value xxxxxxxx  
\v	        -   ASCII Vertical Tab (VT) character  
\oxx        -   Character with octal value xx  
\xhh        -   Character with hex value hh  
</pre>  
  
```python
>>> print("a\tb")
a    b
>>> print("a\141\x61")
aaa
>>> print("a\nb")
a
b
>>> print('\u2192 \N{rightwards arrow}') 
# This type of escape sequence is typically used to insert characters that are not readily generated from the keyboard or are #not easily readable or printable.
```  
#### Raw strings
A raw string literal is preceded by r or R, which specifies that escape sequences in the associated string are not translated. The backslash character is left in the string:   
```python
>>> print('foo\nbar')
foo
bar
>>> print(r'foo\nbar')
foo\nbar

>>> print('foo\\bar')
foo\bar
>>> print(R'foo\\bar')
foo\\bar
```  
**Triple-Quoted Strings**  
```python
>>> print('''This string has a single (') and a double (") quote.''')
This string has a single (') and a double (") quote.
                          
>>> print("""This is a
string that spans
across several lines""")
This is a
string that spans
across several lines
```  
  
**Fundamental Data Types vs Immutability:**  
All Fundamental Data types are immutable. i.e once we creates an object,we cannot
perform any changes in that object. If we are trying to change then with those changes a
new object will be created. This non-chageable behaviour is called immutability.  
  
In Python if a new object is required, then python wont create object immediately. First it
will check if any object available with the required content or not. If available then
existing object will be reused. If it is not available then only a new object will be created.  
  
The advantage of this approach is memory utilization and performance will be improved.
But the problem in this approach is,several references pointing to the same object,by
using one reference if we are allowed to change the content in the existing object then the
remaining references will be effected. To prevent this immutability concept is required.
According to this once creates an object we are not allowed to change content. If we are
trying to change with those changes a new object will be created.  

In [55]:
a = 2
b = 2
print(a == b)
print(a is b)
print(id(a), id(b))

True
True
140723625169760 140723625169760


In [56]:
a = 10 + 5j
b = 10 + 5j
print(a == b)
print(a is b)
print(id(a), id(b))

True
False
2359197449296 2359197650992


In [57]:
a = True
b = True
print(a == b)
print(a is b)
print(id(a), id(b))

True
True
140723624646992 140723624646992


In [58]:
a = 'durga'
b = 'durga'
print(a == b)
print(a is b)
print(id(a), id(b))

True
True
2359197634320 2359197634320


In [62]:
a = 2
b = 2
a = 3 # new object formed
print(a, b)

3 2


**bytes Data Type:**  
- Python supports a range of types to store sequences. There are six sequence types: strings, byte sequences (bytes objects), byte arrays (bytearray objects), lists, tuples, and range objects.  
- bytes data type represents a group of byte numbers just like an array  
- Strings contain Unicode characters. Their literals are written in single or double quotes : 'python', "data". Bytes and bytearray objects contain single bytes 
- Bytes objects can be constructed using constructor, bytes(), and from literals; use a 'b' prefix with normal string syntax: b'python'. To construct byte arrays, use the bytearray() function.  
  
**Create a bytes object in Python**  
```python
>>> x = b"Bytes objects are immutable sequences of single bytes"
>>> print(x)
```

In [63]:
>>> x = b"Bytes objects are immutable sequences of single bytes"
>>> print(x)

b'Bytes objects are immutable sequences of single bytes'


In [64]:
#created from a iterable of ints, string, bytes or buffer objects.
x = bytes('Python, bytes', 'utf8')
print(x)

b'Python, bytes'


**Convert bytes to string**  

In [65]:
x = b'El ni\xc3\xb1o come camar\xc3\xb3n'
print(x)
# create a string using the decode() method of bytes. 
#This method takes an encoding argument, such as UTF-8, and optionally an errors argument.

s = x.decode()
print(type(s))
print(s)

b'El ni\xc3\xb1o come camar\xc3\xb3n'
<class 'str'>
El niño come camarón


In [66]:
#create a bytes object encoded using 'cp855'
x = b'\xd8\xe1\xb7\xeb\xa8\xe5 \xd2\xb7\xe1'
print(x)
#return a string using decode 'cp855'
y = x.decode('cp855')
print(y)

b'\xd8\xe1\xb7\xeb\xa8\xe5 \xd2\xb7\xe1'
привет мир


**Convert hex string to bytes**  

In [71]:
#create a string with hexadecimal data
x = 'B9 01EF'
print(x)
y = bytes.fromhex(x) 
print(y)

B9 01EF
b'\xb9\x01\xef'


**Unicode code**  

In [78]:
x = ord(b'a')
print(x)
print( ord('A') )

97
65


In [79]:
#create a bytes object
y = b'Python bytes'
#generates a list of codes from the characters of bytes
z = list(y)
print(z)

[80, 121, 116, 104, 111, 110, 32, 98, 121, 116, 101, 115]


**Define a mapping table characters for use with a bytes object in Python**  

In [81]:
b_table = bytes.maketrans(b'abcdef', b'uvwxyz')
st = 'Write a Python function to find a distinct pair of numbers whose product is odd from a sequence of integer values.'
b_new = st.translate(b_table)
print(b_new)

Writy u Python zunwtion to zinx u xistinwt puir oz numvyrs whosy proxuwt is oxx zrom u syquynwy oz intygyr vuluys.


**Convert bytes to hex in Python**  

In [84]:
import binascii
a = binascii.hexlify("Python".encode("utf8"))
print(a)
print( binascii.unhexlify(a).decode("utf8") )

b'507974686f6e'
Python


**How to get the character from the numeric code in bytes objects**  

In [87]:
print(chr(60))
print(chr(256))
print(chr(360))
print(chr(1256))

<
Ā
Ũ
Ө


**Create a bytearray object in Python**

In [1]:
x = bytearray("Python Bytes", "utf8")
print(x)
y = bytearray([94, 91, 101, 125, 111, 35, 120, 101, 115, 101, 200])
print(y)

bytearray(b'Python Bytes')
bytearray(b'^[e}o#xese\xc8')


**Difference between bytes and bytearray object in Python**  
- bytes are immutable bytearray are mutuable

In [4]:
x = bytearray("Python bytearray", "utf8")
x.append(45)
print(x)

bytearray(b'Python bytearray-')


In [5]:
x = bytes("Python bytearray", "utf8")
x.append(45)
print(x)

AttributeError: 'bytes' object has no attribute 'append'

**Difference between bytes and string object**    
```python
>>> # bytes objects are immutable sequences of integers, each value in the sequence
>>> # string objects are immutable sequences of unicode characters.
>>> x = "Python String"
>>> y = b"Python String"
>>> print(x)
    Python String
>>> print(y)
    b'Python String'
>>> # Found in unicode representation of characters but not in ascii
>>> x = "Python"
>>> y = bytes("Python", "utf8")
>>> print(x)
    Python
>>> print(y)
    b'Python'
```
**list data type:**  
An ordered, mutable, heterogenous collection of eleemnts is nothing but list, where
duplicates also allowed.  


In [6]:
list=[10,20,30]
list.append("Ravinder")
print(list)
list.remove(20)
print(list)
list2=list*2
print(list2)

[10, 20, 30, 'Ravinder']
[10, 30, 'Ravinder']
[10, 30, 'Ravinder', 10, 30, 'Ravinder']


**tuple data type:**  
- tuple data type is exactly same as list data type except that it is immutable.i.e we cannot
chage values.  
- Tuple elements can be represented within parenthesis.
- tuple is the read only version of list

In [7]:
t=(10,20,30,40)
print(type(t))

t[0]=100

<class 'tuple'>


TypeError: 'tuple' object does not support item assignment

**set Data Type:**  
- If we want to represent a group of values without duplicates where order is not important then we should go for set Data Type.

In [8]:
s={100,0,10,200,10,'durga'}
print(s)

s.add(60)
print(s)

s.remove(100)
print(s)

{0, 100, 200, 10, 'durga'}
{0, 100, 200, 10, 'durga', 60}
{0, 200, 10, 'durga', 60}


**frozenset Data Type**  
- It is exactly same as set except that it is immutable.  
- Hence we cannot use add or remove functions.

In [11]:
s={10,20,30,40}
fs=frozenset(s)
print(type(fs))
fs.add(70)

<class 'frozenset'>


AttributeError: 'frozenset' object has no attribute 'add'

**dict Data Type:**  
- If we want to represent a group of values as key-value pairs then we should go for dict data type.
Eg:
```python
d={101:'durga',102:'ravi',103:'shiva'}
```
- Duplicate keys are not allowed but values can be duplicated. If we are trying
# Operators
Operator is a symbol that performs certain operations.
Python provides the following set of operators
1. Arithmetic Operators
2. Relational Operators or Comparison Operators
3. Logical operators
4. Bitwise oeprators
5. Assignment operators
6. Special operators  
  
### 1. Arithmetic Operators:  
Addition ==> +  
Subtraction ==> -   
Multiplication ==> *    
Division operator ==> /  
Modulo operator ==> %  
Floor Division operator ==> //  
Exponent operator or power operator ==> **  

In [2]:
a=10
b=2
print('a+b =',a+b)
print('a-b =',a-b)
print('a*b =',a*b)
print('a/b =',a/b)
print('a//b =',a//b)
print('a%b =',a%b)
print('a**b =',a**b) 

a+b = 12
a-b = 8
a*b = 20
a/b = 5.0
a//b = 5
a%b = 0
a**b = 100


**NOTE**   
'/' operator always performs floating point arithmetic. Hence it will always returns
float value.  
But Floor division (//) can perform both floating point and integral arithmetic. If
arguments are int type then result is int type. If atleast one argument is float type then
result is float type.
### 2. Relational Operators:
(>,>=,<,<=)  

In [4]:
a=10
b=20
print("a > b is ",a>b)
print("a >= b is ",a>=b)
print("a < b is ",a<b)
print("a <= b is ",a<=b)

a > b is  False
a >= b is  False
a < b is  True
a <= b is  True


In [5]:
a="Ravinder"
b="Ravinder"
print("a > b is ",a>b)
print("a >= b is ",a>=b)
print("a < b is ",a<b) 

a > b is  False
a >= b is  True
a < b is  False


Chaining of relational operators is possible. In the chaining, if all comparisons
returns True then only result is True. If atleast one comparison returns False then the
result is False


In [8]:
print(10<20<30<40>50)
print(10<20<30<40<50)

False
True


### equality operators:
(== , !=)  
We can apply these operators for any type even for incompatible types also

In [10]:
print(10 == 10)
print(10 != 20)
print( 10 != 10 )
print("Ravinder" == 'ravinder')

True
True
False
False


**Note** : Chaining concept is applicable for equality operators. If atleast one comparison
returns False then the result is False. otherwise the result is True.

In [11]:
print(10==20==30==40 )
print(10==10==10==10 )

False
True


### 3. Logical Operators
(and, or ,not)   
and ==> If both arguments are True then only result is True    
or ====> If atleast one arugemnt is True then result is True    
not ==> complement   
  
non-boolean types behaviour:  
- 0 means False
- non-zero means True
- empty string is always treated as False  

**x and y:**  
- if x is evaluates to false return x otherwise return y  

In [33]:
print(9 and 7) #if x is evaluates to false return x otherwise return y

7


In [34]:
x = True
y = False

# Output: x and y is False
print('x and y is',x and y)

# Output: x or y is True
print('x or y is',x or y)

# Output: not x is False
print('not x is',not x)

x and y is False
x or y is True
not x is False


**x or y:**  
If x evaluates to True then result is x otherwise result is y  

In [35]:
print(10 or 20)
print(0 or 20)

10
20


**not x:**  
- If x is evalutates to False then result is True otherwise False

In [36]:
print(not 10)
print(not 0)

False
True


In [41]:
print("KIKI" and "ravi")
print("" and "KIKI")
print("KIKI" and "")
print("" or "KIKI")
print("KIKI" or "")
print(not "")
print(not "KIKI")

ravi


KIKI
KIKI
True
False


### 4. Bitwise Operators
- We can apply these operators bitwise.
- These operators are applicable only for int and boolean types.
- By mistake if we are trying to apply for any other type then we will get Error  
(&,|,^,~,<<,>>)  
- Bitwise operators act on operands as if they were string of binary digits. It operates bit by bit, hence the name.
- In the table below: Let x = 10 (0000 1010 in binary) and y = 4 (0000 0100 in binary)   
  
| Operator | Meaning                                                           | Example                |
|----------|-------------------------------------------------------------------|------------------------|
| &        | If both bits are 1 then only result is 1 otherwise result is 0    | x & y = 0 (0000 0000)  |
| |        | If atleast one bit is 1 then result is 1 otherwise result is 0    | x | y = 14 (0000 1110) |
| ~        |  bitwise complement operator i.e 1 means 0 and 0 means 1          | ~x = -11 (1111 0101)   |
| ^        | If bits are different then only result is 1 otherwise result is 0 | x ^ y = 14 (0000 1110) |
| >>       | Bitwise right shift                                               | x>> 2 = 2 (0000 0010)  |
| <<       | Bitwise left shift                                                | x<< 2 = 40 (0010 1000) |  
### 5. Assignmnet Operators
- Assignment operators are used in Python to assign values to variables.

- a = 5 is a simple assignment operator that assigns the value 5 on the right to the variable a on the left.

- There are various compound operators in Python like a += 5 that adds to the variable and later assigns the same. It is equivalent to a = a + 5.  
  
| Operator | Exmple  | Equivalent to |
|----------|---------|---------------|
| =        | x = 5   | x = 5         |
| +=       | x += 5  | x = x + 5     |
| -=       | x -= 5  | x = x - 5     |
| *=       | x *= 5  | x = x * 5     |
| /=       | x /= 5  | x = x / 5     |
| %=       | x %= 5  | x = x % 5     |  
| //=      | x //= 5 | x = x // 5    |
| **=      | x **= 5 | x = x ** 5    |
| &=       | x &= 5  | x = x & 5     |
| |=       | x |= 5  | x = x | 5     |
| ^=       | x ^= 5  | x = x ^ 5     |
| >>=      | x >>= 5 | x = x >> 5    |
| <<=      | x <<= 5 | x = x << 5    |  
  
### 6. Ternary Operator:  
x = firstValue if condition is True else secondValue  
```python
a if condition else b
```  
Ternory operators can be nested

In [2]:
a = 30
b = 20
print('a is greater than b') if(a > b) else print('b is greater than a')

a is greater than b


**WAP to find minimum of 3 numbers**  

In [22]:
a = int(input('Enter First number...'))
b = int(input('Enter Second number...'))
c = int(input('Enter Third number...'))

min = a if((a<b) and (a<c)) else b if(b<c) else c
print(min)

Enter First number...3
Enter Second number...4
Enter Third number...2
2


**WAP to find maximum of 3 numbers**  

In [21]:
a = int(input('Enter First number...'))
b = int(input('Enter Second number...'))
c = int(input('Enter Third number...'))

max = a if((a>b) and (a>c)) else b if(b>c) else c
print(max)

Enter First number...3
Enter Second number...4
Enter Third number...2
2


In [7]:
a = 3
b = 2
c = 1
if((a>b) and (a>c)):
    print(a)
elif b>c:
    print(b)
else:
    print(c)

3


In [8]:
a = int(input("Enter First Number:"))
b = int(input("Enter Second Number:"))
print("Both numbers are equal" if a==b else "First Number is Less than Second Number" if a<b else "First Number Greater than Second Number") 

Enter First Number:2
Enter Second Number:3
First Number is Less than Second Number


### 7. Special operators:
Python language offers some special type of operators like the identity operator or the membership operator. They are described below with examples.  
**Identity operators(address comparison)**  
is and is not are the identity operators in Python. They are used to check if two values (or variables) are located on the same part of the memory. Two variables that are equal does not imply that they are identical.  
```python
r1 is r2 returns True if both r1 and r2 are pointing to the same object
r1 is not r2 returns True if both r1 and r2 are not pointing to the same object
```


In [13]:
x1 = 5
y1 = 5
x2 = 'Hello'
y2 = 'Hello'
x3 = [1,2,3]
y3 = [1,2,3]

# Output: False
print(x1 is not y1)
print(id(x1))
print(id(y1))

# Output: True
print(x2 is y2)
print(id(x2))
print(id(y2))

# Output: False
print(x3 is y3)
print(id(x3))
print(id(y3))

False
140736679809984
140736679809984
True
2121191551976
2121191551976
False
2121191606024
2121191605960


Here, we see that x1 and y1 are integers of same values, so they are equal as well as identical. Same is the case with x2 and y2 (strings).  

But x3 and y3 are list. They are equal but not identical. It is because interpreter locates them separately in memory although they are equal.

In [16]:
x3 = [1,2,3]
y3 = x3
print(x3 is y3)

True


- Immutable Object: int, float, long, complex, string tuple, bool
- Mutable Object: list, dict, set, byte array, user-defined classes  
  
Let’s see what happens when we apply id() on an immutable object, an integer:  
  
```python
>>> a = 89
>>> id(a)
140736679812672
>>> a = 89 + 1
>>> print(a)
90
>>> id(a)
140736679812704  # this is different from before!
```  
  
…and contrasting the result with a mutable object, a list:  
```python
>>> L = [1, 2, 3]
>>> id(L)
2121190586120
>>> L += [4]
>>> print(L)
[1, 2, 3, 4]
>>> id(L)
2121190586120    # this is the same as before! 
```

We see that when we attempt to modify an immutable object (integer in this case), Python simply gives us a different object instead. On the other hand, we are able to make changes to an mutable object (a list) and have it remain the same object throughout.   
  
It’s important to distinguish the identity function id() and identity operator is from the comparison operator ==, which evaluates whether the values are equal.  
**'==' VS 'is' operators**  
Read this [article](http://foobarnbaz.com/2012/07/08/understanding-python-variables/)  
For string interning read this [article](http://guilload.com/python-string-interning/)  
**Membership operators:**  
in and not in are the membership operators in Python. They are used to test whether a value or variable is found in a sequence (string, list, tuple, set and dictionary).  

In a dictionary we can only test for presence of key, not the value.  
- in --> Returns True if the given object present in the specified Collection
- not in --> Retruns True if the given object not present in the specified Collection

In [24]:
x = 'Hello world'
y = {1:'a',2:'b'}

# Output: True
print('H' in x)

# Output: True
print('hello' not in x)

# Output: True
print(1 in y)

# Output: False
print('a' in y)

True
True
True
False


### Operator Precedence:  
If multiple operators present then which operator will be evaluated first is decided by
operator precedence.  
  
| Operator           | Name                                             |
|--------------------|--------------------------------------------------|
| ()                 | Parenthesis                                      |
| **                 | exponential                                      |
| ~,-                | Bitwise complement operator,unary minus operator |
| *,/,%,//           | multiplication,division,modulo,floor division    |
| +,-                | addition,subtraction                             |
| <<,>>              | Left and Right Shift                             |
| &                  | bitwise And                                      |
| ^                  | Bitwise X-OR                                     |
| |                  | Bitwise OR                                       |
| >,>=,<,<=, ==, !=  | Relational or Comparison operators               |
| =,+=,-=,*=...      | Assignment operators                             |
| is , is not        | Identity Operators                               |
| in , not in        | Membership operators                             |
| not                | Logical not                                      |
| and                | Logical and                                      |
| or                 | Logical or                                       |

In [25]:
a=30
b=20
c=10
d=5
print((a+b)*c/d) #100.0
print((a+b)*(c/d)) #100.0
print(a+(b*c)/d) #70.0 

100.0
100.0
70.0


# Modules  
**Mathematical Functions (math Module)**  
- A Module is collection of functions, variables and classes etc.
- math is a module that contains several functions to perform mathematical operations
- If we want to use any module in Python, first we have to import that module.
```python
import math
```  
  
Once we import a module then we can call any function of that module

In [1]:
import math
print(math.sqrt(16))
print(math.pi)

4.0
3.141592653589793


**We can create alias name by using as keyword.**  
```python
import math as m
```

In [3]:
import math as m
print(m.sqrt(16))
print(m.pi)

4.0
3.141592653589793


**We can import a particular member of a module explicitly as follows:**  
```python
from math import sqrt
from math import sqrt,pi
```

In [4]:
from math import sqrt,pi
print(sqrt(16))
print(pi)
print(math.pi)

4.0
3.141592653589793
3.141592653589793


important functions of math module:  
ceil(x)  
floor(x)  
pow(x,y)  
factorial(x)  
trunc(x)  
gcd(x,y)  
sin(x)  
cos(x)   
tan(x)  
  
**important variables of math module:**  
  
| varibale           | value                              |
|--------------------|------------------------------------|
| e                  | 2.71                               |
| inf                | infinity                           |
| nan                | not a number                       |  
  
**Write a Python program to find area of circle**  


In [5]:
from math import pi
r=16
print("Area of Circle is :",pi*r**2)

Area of Circle is : 804.247719318987


# Input And Output Statements
- input() function can be used to read data directly in our required format.We are not required to perform type casting.
```python
x=input("Enter Value)
print(type(x))
```  
  
every input value is treated as str type only.  
**Write a program to read 2 numbers from the keyboard and print sum.**  

In [6]:
x=input("Enter First Number:")
y=input("Enter Second Number:")
i = int(x)
j = int(y)
print("The Sum:",i+j) 

Enter First Number:5
Enter Second Number:6
The Sum: 11


In [7]:
x=int(input("Enter First Number:"))
y=int(input("Enter Second Number:"))
print("The Sum:",x+y) 

Enter First Number:5
Enter Second Number:6
The Sum: 11


**How to read multiple values from the keyboard in a single line:**  
    

In [11]:
a,b= [int(x) for x in input("Enter 2 numbers :").split()]
print("Product is :", a*b) 

Enter 2 numbers :5 4
Product is : 20


**Q. Write a program to read 3 float numbers from the keyboard with , seperator and print
their sum.**

In [13]:
a,b,c= [float(x) for x in input("Enter 3 float numbers :").split(',')]
print("The Sum is :", a+b+c) 

Enter 3 float numbers :4,5,6
The Sum is : 15.0


**eval():** 

In [15]:
x = eval("10+20+30")
print(x)

60


In [18]:
x = eval(input("Enter Expression"))
print(x)

Enter Expression10+2*3/4
11.5


eval() can evaluate the Input to list, tuple, set, etc based the provided Input.

In [21]:
l = eval(input("Enter List"))
print (type(l))
print(l) 

Enter List1,2,3
<class 'tuple'>
(1, 2, 3)


# Command Line Arguments
The Argument which are passing at the time of execution are called Command Line
Arguments  
  
Till now, we have taken input in python using raw_input() or input(). There is another method that uses command line arguments. The command line arguments must be given whenever we want to give the input before the start of the script, while on the other hand, raw_input() is used to get the input while the python program / script is running.  

For example, in the UNIX environment, the arguments ‘-a’ and ‘-l’ for the ‘ls’ command give different results.  

The command line arguments in python can be processed by using either ‘sys’ module or ‘argparse’ module.  
- argv is not Array it is a List. It is available in sys Module.
- The Argument which are passing at the time of execution are called Command Line Arguments.


In [ ]:
from sys import argv
print("The Number of Command Line Arguments:", len(argv))
print("The List of Command Line Arguments:", argv)
print("Command Line Arguments one by one:")
for x in argv:
    print(x) 

![](img\cml1.png)

In [ ]:
from sys import argv
sum=0
args=argv[1:]
for x in args :
    n=int(x)
    sum=sum+n
print("The Sum:",sum) 

![](img\cml2.png)  
  
**Note1:** usually space is seperator between command line arguments. If our command line
argument itself contains space then we should enclose within double quotes(but not
single quotes)


In [ ]:
from sys import argv
print(argv[1]) 

![](img\cml3.png)  
  
**Note2:** Within the Python program command line arguments are available in the String
form. Based on our requirement,we can convert into corresponding type by using type
casting methods.  
  
**Note3:** If we are trying to access command line arguments with out of range index then
we will get Error.  
  
# output statements
- We can use print() function to display output.  
- print() without any argument, Just it prints new line character  
  

In [29]:
print("Hello World")

#We can use escape characters also
print("Hello \n World")
print("Hello\tWorld")

#We can use repetetion operator (*) in the string
print(10*"Hello")
print("Hello"*10)

#We can use + operator also
print("Hello"+"World") 

Hello World
Hello 
 World
Hello	World
HelloHelloHelloHelloHelloHelloHelloHelloHelloHello
HelloHelloHelloHelloHelloHelloHelloHelloHelloHello
HelloWorld


Note:  
- If both arguments are String type then + operator acts as concatenation operator.
- If one argument is string type and second is any other type like int then we will get Error
- If both arguments are number type then + operator acts as arithmetic addition operator.  


In [30]:
print("Hello"+"World")
print("Hello","World") 

HelloWorld
Hello World


**print with variable number of arguments:**

In [32]:
a,b,c=10,20,30
print("The Values are :",a,b,c) 

The Values are : 10 20 30


By default output values are seperated by space.If we want we can specify seperator by
using "sep" attribute.

In [33]:
a,b,c=10,20,30
print(a,b,c,sep=',')
print(a,b,c,sep=':') 

10,20,30
10:20:30


**print() with end attribute:**

In [35]:
print("Hello")
print("Hi")
print("Yo") 

print("Hello",end=' ')
print("Hi",end=' ')
print("Yo") 

Hello
Hi
Yo
Hello Hi Yo


**Note:** The default value for end attribute is \n,which is nothing but new line character.  
**print(object) statement:**  
We can pass any object (like list,tuple,set etc)as argument to the print() statement. 

In [36]:
l=[10,20,30,40]
t=(10,20,30,40)
print(l)
print(t) 

[10, 20, 30, 40]
(10, 20, 30, 40)


**print(String,variable list):**  
We can use print() statement with String and any number of arguments.  


In [38]:
s="Ravinder"
a=23
s1="C++"
s2="Python"
print("Hello",s,",Your Age is",a)
print("You are teaching",s1,"and",s2) 

Hello Ravinder ,Your Age is 23
You are teaching C++ and Python


**print(formatted string):**  
%i====>int  
%d====>int  
%f=====>float  
%s======>String type   
**print("formatted string" %(variable list))**  

In [39]:
a=10
b=20
c=30
print("a value is %i" %a)
print("b value is %d and c value is %d" %(b,c)) 

a value is 10
b value is 20 and c value is 30


In [40]:
s="Durga"
list=[10,20,30,40]
print("Hello %s ...The List of Items are %s" %(s,list)) 

Hello Durga ...The List of Items are [10, 20, 30, 40]


 **print() with replacement operator {}**   

In [42]:
name="Ravinder"
salary=10000
gf="Rohit"
print("Hello {0} your salary is {1} and Your Friend {2} is waiting".format(name, salary, gf))
print("Hello {x} your salary is {y} and Your Friend {z} is waiting".format(x=name, y=salary, z=gf))

Hello Ravinder your salary is 10000 and Your Friend Rohit is waiting
Hello Ravinder your salary is 10000 and Your Friend Rohit is waiting


# Flow Control
Flow control describes the order in which statements will be executed at runtime.  
![](img\sav.png)  
### Conditional Statements
**if**  

```python
if condition : statement  
or 
if condition :  
    statement-1  
    statement-2  
    statement-3  
If condition is true then statements will be executed.  
```

In [7]:
name=input("Enter Your Name:")
if name=="Ravinder":
    print("Hello %s, Good Morning"%name)
print("How are you!!!") 

Enter Your Name:Ravinder
Hello Ravinder, Good Morning
How are you!!!


**if-else:**  
```python
if condition :
    Action-1
else:
    Action-2
```  
  
if condition is true then Action-1 will be executed otherwise Action-2 will be executed

In [8]:
name=input("Enter Name:")
if name=="Ravinder" :
    print("Hello Ravinder, Good Morning")
else:
    print("Hello, Who are you?")
print("How are you!!!") 

Enter Name:Ravinder
Hello Ravinder, Good Morning
How are you!!!


### if-elif-else:
Syntax:  
```python
if condition1:
    Action-1
elif condition2:
    Action-2
elif condition3:
    Action-3
elif condition4:
    Action-4
...
else:
    Default Action
```  
Based condition the corresponding action will be executed.

In [9]:
brand = input("Enter your favourite brand : ")
if brand == 'Addidas':
    print('It\'s shoes brand')
elif brand == 'Puma':
    print('It\'s another shoe brand')
else:
    print('You chose an unkown brand')

Enter your favourite brand : Puma
It's another shoe brand


**There is no switch statement in Python**  

In [ ]:
from pyswitch import Switch   # pyswitch can be found on PyPI

myswitch = Switch()

@myswitch.case(42)
def case42(value):
    print "I got 42!"

@myswitch.case(range(10))
def caseRange10(value):
    print "I got a number from 0-9, and it was %d!" % value

@myswitch.caseIn('lo')
def caseLo(value):
    print "I got a string with 'lo' in it; it was '%s'" % value

@myswitch.caseRegEx(r'\b([Pp]y\w)\b')
def caseReExPy(matchOb):
    print r"I got a string that matched the regex '\b[Pp]y\w\b', and the match was '%s'" % matchOb.group(1)

@myswitch.default
def caseDefault(value):
    print "Hey, default handler here, with a value of %r." % value

myswitch(5)  # prints: I got a number from 0-9, and it was 5!
myswitch('foobar')  # prints: Hey, default handler here, with a value of foobar.
myswitch('The word is Python')  # prints: I got a string that matched the regex '\b[Pp]y\w\b', and the match was 'Python'

In [25]:
def switch(x):
    return {
        1 : 'output for case 1',
        2 : 'output for case 2',
        3 : 'output for case 3'
    }.get(x, 'default case')  

In [26]:
f(5)

'default case'

In [36]:
choices = {'a': 'output for case a', 'b': 'output for case b'}
#choices.get('key_name', 'default')
choices.get('b', 'default')

'output for case b'

```C
// C Language version of a simple 'switch/case'.
switch( key ) 
{
    case 'a' :
        print('output for case 1')
        break;
    case 'b' :
        print('output for case 1')
        break;
    default :
        print('default')
}
```

**Write a program to take a single digit number from the key board and print is value in English word?**  

In [37]:
n=int(input("Enter a digit from o to 9:"))
if n==0 :
    print("ZERO")
elif n==1:
    print("ONE")
elif n==2:
    print("TWO")
elif n==3:
    print("THREE")
elif n==4:
    print("FOUR")
elif n==5:
    print("FIVE")
elif n==6:
    print("SIX")
elif n==7:
    print("SEVEN")
elif n==8:
    print("EIGHT")
elif n==9:
    print("NINE")
else:
    print("PLEASE ENTER A DIGIT FROM 0 TO 9") 

Enter a digit from o to 9:7
SEVEN


### Iterative Statements
- If we want to execute a group of statements multiple times then we should go for Iterative statements.
- Python supports 2 types of iterative statements.
- **for loop**
- **while loop**  
  
**for loop:**  
- If we want to execute some action for every element present in some sequence(it may be string or collection)then we should go for 'for' loop.
*Syntax:*   
```python
for x in sequence :
    body
```
- where sequence can be string or any collection.
- Body will be executed for every element present in the sequence  
  
**To print numbers from 1 to 10 by using while loop**  

In [43]:
for i in range(1, 11):
    print(i)

1
2
3
4
5
6
7
8
9
10


**To print characters present in string index wise:**  

In [40]:
s=input("Enter Your Name: ")
i=0
for x in s :
    print("The character present at ",i,"index is :",x)
    i=i+1 

Enter Your Name: Ravinder Singh
The character present at  0 index is : R
The character present at  1 index is : a
The character present at  2 index is : v
The character present at  3 index is : i
The character present at  4 index is : n
The character present at  5 index is : d
The character present at  6 index is : e
The character present at  7 index is : r
The character present at  8 index is :  
The character present at  9 index is : S
The character present at  10 index is : i
The character present at  11 index is : n
The character present at  12 index is : g
The character present at  13 index is : h


In [41]:
s=input("Enter Your Name: ")
for i, x in enumerate(s):
    print("The character present at ",i,"index is :",x)

Enter Your Name: Ravi Singh
The character present at  0 index is : R
The character present at  1 index is : a
The character present at  2 index is : v
The character present at  3 index is : i
The character present at  4 index is :  
The character present at  5 index is : S
The character present at  6 index is : i
The character present at  7 index is : n
The character present at  8 index is : g
The character present at  9 index is : h


**while loop:**  
If we want to execute a group of statements iteratively until some condition false,then we
should go for while loop.  
*Syntax:*  
```python
while condition :
    body
```  
**To print numbers from 1 to 10 by using while loop**  

In [42]:
x=1
while x <=10:
    print(x)
    x=x+1 

1
2
3
4
5
6
7
8
9
10


**To display the sum of first n numbers**  

In [47]:
n = int(input('Enter value of n : '))
sum = 0
i = 1
while(i < n+1):
    sum += i
    i += 1
print(sum)

Enter value of n : 5
15


**Nested Loops:**  
Sometimes we can take a loop inside another loop,which are also known as nested loops.  

In [1]:
for i in range(4):
    for j in range(4):
        print("i=",i," j=",j) 

i= 0  j= 0
i= 0  j= 1
i= 0  j= 2
i= 0  j= 3
i= 1  j= 0
i= 1  j= 1
i= 1  j= 2
i= 1  j= 3
i= 2  j= 0
i= 2  j= 1
i= 2  j= 2
i= 2  j= 3
i= 3  j= 0
i= 3  j= 1
i= 3  j= 2
i= 3  j= 3


**Write a program to dispaly *'s in Right angled triangled form**    
<pre>
*
* *
* * *
* * * *
* * * * *
* * * * * *
* * * * * * *
</pre>

In [5]:
n = int(input("Enter number of rows:")) 
for row in range(1, n+1):
    for col in range(1, row+1):
        print('*', end = " ")
    print()

Enter number of rows:7
* 
* * 
* * * 
* * * * 
* * * * * 
* * * * * * 
* * * * * * * 


In [14]:
n = int(input("Enter number of rows:"))
for i in range(1,n+1):
    print('* '*i) 

Enter number of rows:7
* 
* * 
* * * 
* * * * 
* * * * * 
* * * * * * 
* * * * * * * 


** Write a program to display *'s in pyramid style(also known as equivalent triangle)**  
<pre>
      *
     * *
    * * *
   * * * *
  * * * * *
 * * * * * *
* * * * * * *  
</pre>

In [3]:
n = int(input("Enter number of rows:")) 
for row in range(1, n+1):
    k=0
    for space in range(1, n-row+1):
        print(' ', end = " ")
    while(k != 2*row-1):
        print('*', end = " ")
        k+=1
    print()

Enter number of rows:7
            * 
          * * * 
        * * * * * 
      * * * * * * * 
    * * * * * * * * * 
  * * * * * * * * * * * 
* * * * * * * * * * * * * 


In [4]:
n = int(input("Enter number of rows:"))
for i in range(1,n+1):
    print(" " * (n-i),end="")
    print("* "*i) 

Enter number of rows:7
      * 
     * * 
    * * * 
   * * * * 
  * * * * * 
 * * * * * * 
* * * * * * * 


### Transfer Statements/ Jumping Statements
**break:**  
We can use break statement inside loops to break loop execution based on some
condition.  

In [8]:
for i in range(10):
    if i==7:
        print("value of i = 7, break")
        break
    print(i) 

0
1
2
3
4
5
6
value of i = 7, break


In [9]:
cart=[10,20,600,60,70]
for item in cart:
    if item>500:
        print("To place this order insurence must be required")
        break
    print(item) 

10
20
To place this order insurence must be required


**continue:**  
We can use continue statement to skip current iteration and continue next iteration.  

In [10]:
for i in range(10):
    if i%2==0:
        continue
    print(i) 

1
3
5
7
9


In [11]:
cart=[10,20,500,700,50,60]
for item in cart:
    if item>=500:
        print("We cannot process this item :",item)
        continue
    print(item) 

10
20
We cannot process this item : 500
We cannot process this item : 700
50
60


**What is the difference between for loop and while loop in Python?**  
- Repeat code for every item in sequence ==>for loop
- Repeat code as long as condition is true ==>while loop  
  
**How to exit from the loop?**  
- by using break statement  
  
**How to skip some iterations inside loop?**  
- by using continue statement  
  
**When else part will be executed wrt loops?**  
- If loop executed without break  
  
**pass statement:**  
- pass is a keyword in Python.
- In our programming syntactically if block is required which won't do anything then we can
define that empty block with pass keyword.
- It is an empty statement
- It is null statement
- It won't do anything

In [12]:
def m1(): 
    #do nothing
    pass

**use case of pass:**  
Sometimes in the parent class we have to declare a function with empty body and child
class responsible to provide proper implementation. Such type of empty body we can
define by using pass keyword. (It is something like abstract method in java)

In [13]:
for i in range(100):
    if i%9==0:
        print(i)
    else:
        pass

0
9
18
27
36
45
54
63
72
81
90
99


**del statement:**  
- del is a keyword in Python.
- After using a variable, it is highly recommended to delete that variable if it is no longer
required,so that the corresponding object is eligible for Garbage Collection.
- We can delete variable by using del keyword.  

In [14]:
x=10
print(x)
del x 
print(x)

10


NameError: name 'x' is not defined

**Note:**  
We can delete variables which are pointing to immutable objects.But we cannot delete
the elements present inside immutable object.

In [15]:
s="durga"
print(s)
del s # valid
s="durga"
print(s)
del s[0] # invalid

durga
durga


TypeError: 'str' object doesn't support item deletion

**Difference between del and None:**  
- In the case del, the variable will be removed and we cannot access that variable(unbind
operation)  
- But in the case of None assignment the variable won't be removed but the corresponding
object is eligible for Garbage Collection(re bind operation). Hence after assigning with
None value,we can access that variable.  
# String
- The most commonly used object in any project and in any programming language is String only.
- Hence we should aware complete information about String data type.  
- This tutorial focuses on common problems involving text
manipulation, such as pulling apart strings, searching, substitution, lexing, and parsing.
- Many of these tasks can be easily solved using built-in methods of strings. However,
more complicated operations might require the use of regular expressions or the cre‐
ation of a full-fledged parser. All of these topics are covered.   
> Any sequence of characters within either single quotes or double quotes is considered as a String.  


In [16]:
s1='durga'
s2="durga"
print(type(s1))
print(type(s2))

<class 'str'>
<class 'str'>


**Note:** In most of other languges like C, C++,Java, a single character with in single quotes is treated
as char data type value. But in Python we are not having char data type.Hence it is treated as
String only

**define multi-line String literals:**  
We can define multi-line String literals by using triple single or double quotes.  
Eg:  
```python
>>> s='''I'm 
Ravinder 
singh  
```  
  
**How to access characters of a String:**  
We can access characters of a string by using the following ways.  
1. By using index  
2. By using slice operator   
**1. By using index:**  
- Python supports both +ve and -ve index.
- positive index means left to right(Forward direction)
- negative index means right to left(Backward direction)

In [20]:
s = 'Ravi'
print(s[0], s[1], s[2], s[3])
print(s[-1],s[-2], s[-3], s[-4])

R a v i
i v a R


**Note:** If we are trying to access characters of a string with out of range index then we will get
error saying : IndexError  
**Write a program to accept some string from the keyboard and display its characters by
index wise(both positive and nEgative index)**  

In [22]:
s=input("Enter Some String:")
i=0
for x in s:
    print("The character present at Positive index {} and at Negative index {} is {}".format(i,i-len(s),x))
    i=i+1 

Enter Some String:RAVI
The character present at Positive index 0 and at Negative index -4 is R
The character present at Positive index 1 and at Negative index -3 is A
The character present at Positive index 2 and at Negative index -2 is V
The character present at Positive index 3 and at Negative index -1 is I


**Accessing characters by using slice operator:**  
```python
str[Begin_index:end_index:step]
```  
  
**Note:** If we are not specifying begin index then it will consider from beginning of the string.
If we are not specifying end index then it will consider up to end of the string
The default value for step is 1


In [6]:
s="Learning Python is very very easy!!!"
print(s[0:8:1])
print(s[9::1])
print(s[0:8:2])
print(s[::-1])

Learning
Python is very very easy!!!
Lann
!!!ysae yrev yrev si nohtyP gninraeL


**Behaviour of slice operator:**  
  
- step value can be either +ve or –ve  
- if +ve then it should be forward direction(left to right) and we have to consider bEgin to end-1   
- if -ve then it should be backward direction(right to left) and we have to consider bEgin to end+1  
  
**Mathematical Operators for String:**  
We can apply the following mathematical operators for Strings.
1. + operator for concatenation
2. * operator for repetition

Note:
1. To use + operator for Strings, compulsory both arguments should be str type
2. To use * operator for Strings, compulsory one argument should be str and other argument should be int

In [8]:
print("Ravinder" + " " + "Singh") 
print("Ravinder "*2) 

Ravinder Singh
Ravinder Ravinder 


**len() in-built function:**  
We can use len() function to find the number of characters present in the string.

In [9]:
print(len(s))

36


**Checking Membership:**  
We can check whether the character or string is the member of another string or not by using in
and not in operators


In [10]:
s='Ravinder'
print('d' in s) 
print('z' in s) 

True
False


**Comparison of Strings:**  
We can use comparison operators (<,<=,>,>=) and equality operators(==,!=) for strings.
Comparison will be performed based on alphabetical order.     

In [12]:
s1=input("Enter first string:")
s2=input("Enter Second string:")
if s1==s2:
    print("Both strings are equal")
elif s1<s2:
    print("First String is less than Second String")
else:
    print("First String is greater than Second String")

Enter first string:Ravinder
Enter Second string:Ravinder
Both strings are equal


**Removing spaces from the string:**  
We can use the following 3 methods:  
   
| functions | Descrip                             |
|-----------|-------------------------------------|
| rstrip()  | To remove spaces at right hand side |
| strip()   | To remove spaces at left hand side  |
| lstrip()  | To remove spaces both sides         |   
  

In [5]:
st = '  Hello   World!  '
print(st.strip())
print(st.lstrip())
print(st.rstrip())

Hello   World!
Hello   World!  
  Hello   World!


### Finding Substrings:
We can use the following 4 methods  
**For forward direction:**  
- find()
- index()  
**For backward direction:**  
- rfind()
- rindex()  
  
**1. find():**  
```python
s.find(substring)  
```
Returns index of first occurrence of the given substring. If it is not available then we will get -1


In [6]:
s="Learning Python is very easy" 
print(s.find("Python")) #9
print(s.find("Java")) # -1
print(s.find("r"))#3
print(s.rfind("r"))#21 

9
-1
3
21


**Note:** By default find() method can search total string. We can also specify the boundaries to
search.  
```python
s.find(substring,begin,end)
```  
It will always search from bEgin index to end-1 index

In [8]:
s="RavinderSingh"
print(s.find('i'))#3
print(s.find('i',4,12))#9
print(s.find('z',7,15))#-1 

3
9
-1


**index() method:**  
index() method is exactly same as find() method except that if the specified substring is not
available then we will get ValueError.  


In [9]:
s=input("Enter main string:")
subs=input("Enter sub string:")
try:
    n=s.index(subs)
except ValueError:
    print("substring not found")
else:
    print("substring found")

Enter main string:learning python is very easy
Enter sub string:python
substring found


**Program to display all positions of substring in a given main string**  

In [11]:
s=input("Enter main string:")
subs=input("Enter sub string:")

flag=False
pos=-1
n=len(s)

while True:
    pos=s.find(subs,pos+1,n)
    if pos==-1:
        break
    print("Found at position",pos)
    flag=True
if flag==False:
    print("Not Found") 

Enter main string:abbababababbacdefgbb
Enter sub string:bb
Found at position 1
Found at position 10
Found at position 18


**Counting substring in the given String:**  
We can find the number of occurrences of substring present in the given string by using count()
method.  
1. s.count(substring) ==> It will search through out the string
2. s.count(substring, bEgin, end) ===> It will search from bEgin index to end-1 index

In [12]:
s="abcabcabcabcadda"
print(s.count('a'))
print(s.count('ab'))
print(s.count('a',3,7)) 

6
4
2


**Replacing a string with another string:**  
```python
s.replace(oldstring,newstring)
```
inside s, every occurrence of oldstring will be replaced with newstring  

In [13]:
s="Learning Python is very difficult"
s1=s.replace("difficult","easy")
print(s1)

Learning Python is very easy


In [13]:
s="ababababababab"
s1=s.replace("a","b")
print(s1)

bbbbbbbbbbbbbb


**Q. String objects are immutable then how we can change the content by
using replace() method.**  
- Once we creates string object, we cannot change the content.This non changeable behaviour is
nothing but immutability. If we are trying to change the content by using any method, then with
those changes a new object will be created and changes won't be happend in existing object.
- Hence with replace() method also a new object got created but existing object won't be changed.

In [15]:
s="abab"
print(s,"is available at :",id(s))
s=s.replace("a","b")
print(s1,"is available at :",id(s1))

abab is available at : 2474481223696
bbbb is available at : 2474481224592


**Splitting of Strings:**  
- We can split the given string according to specified seperator by using split() method.
l=s.split(seperator)  
- The default seperator is space. The return type of split() method is List


In [18]:
s="Ravinder Singh Tutorials"
l=s.split()
for x in l:
    print(x) 

Ravinder
Singh
Tutorials


In [19]:
s="23-04-109"
l=s.split('-')
for x in l:
    print(x)

23
04
109


**Joining of Strings**  
- We can join a group of strings(list or tuple) wrt the given seperator.
```python
s=seperator.join(group of strings)
```

In [20]:
t=('sunny','bunny','money')
s=' '.join(t)
print(s)

sunny bunny money


In [21]:
l=['hyderabad','singapore','london','dubai']
s=', '.join(l)
print(s)

hyderabad, singapore, london, dubai


**Changing case of a String:**  
We can change case of a string by using the following 4 methods.
1. **upper()**  ===>To convert all characters to upper case
2. **lower()**  ===>To convert all characters to lower case
3. **swapcase()**  ===>converts all lower case characters to upper case and all upper case characters to
lower case
4. **title()**  ===>To convert all character to title case. i.e first character in every word should be upper
case and all remaining characters should be in lower case.
5. **capitalize()**  ==>Only first character will be converted to upper case and all remaining characters
can be converted to lower case 


In [22]:
s='learning Python is very Easy'
print(s.upper())
print(s.lower())
print(s.swapcase())
print(s.title())
print(s.capitalize())

LEARNING PYTHON IS VERY EASY
learning python is very easy
LEARNING pYTHON IS VERY eASY
Learning Python Is Very Easy
Learning python is very easy


**Checking starting and ending part of the string:**  
Python contains the following methods for this purpose
1. s.startswith(substring)
2. s.endswith(substring)

In [23]:
s='learning Python is very easy'
print(s.startswith('learning'))
print(s.endswith('learning'))
print(s.endswith('easy'))

True
False
True


**To check type of characters present in a string:**  
Python contains the following methods for this purpose.  
1. **isalnum():** Returns True if all characters are alphanumeric( a to z , A to Z ,0 to9 )  
2. **isalpha():** Returns True if all characters are only alphabet symbols(a to z,A to Z)  
3. **isdigit():** Returns True if all characters are digits only( 0 to 9)  
4. **islower():** Returns True if all characters are lower case alphabet symbols  
5. **isupper():** Returns True if all characters are upper case aplhabet symbols  
6. **istitle():** Returns True if string is in title case  
7. **isspace():** Returns True if string contains only spaces  

In [25]:
print('Ravi9582'.isalnum()) #True
print('ravi786'.isalpha()) #False
print('ravi'.isalpha()) #True
print('ravi'.isdigit()) #False
print('786786'.isdigit()) #True
print('abc'.islower()) #True
print('Abc'.islower()) #False
print('abc123'.islower()) #True
print('ABC'.isupper()) #True
print('Learning python is Easy'.istitle()) #False
print('Learning Python Is Easy'.istitle()) #True
print(' '.isspace()) #True

True
False
True
False
True
True
False
True
True
False
True
True


**Formatting the Strings:**  
We can format the strings with variable values by using replacement operator {} and format()
method.  
**Case- 1: Basic Formatting for default, positional and keyword arguments**

In [29]:
name='Ravi'
salary=100000
age=50
print("{}'s salary is {} and his age is {}".format(name,salary,age))
print("{0}'s salary is {1} and his age is {2}".format(name,salary,age))
print("{0}'s age is {2} and his salary is {1}".format(name,salary,age))
print("{x}'s salary is {y} and his age is {z}".format(z=age,y=salary,x=name))

Ravi's salary is 100000 and his age is 50
Ravi's salary is 100000 and his age is 50
Ravi's age is 50 and his salary is 100000
Ravi's salary is 100000 and his age is 50


**Case-2: Formatting Numbers**  
d--->Decimal IntEger  
f----->Fixed point number(float).The default precision is 6  
b-->Binary format  
o--->Octal Format  
x-->Hexa Decimal Format(Lower case)  
X-->Hexa Decimal Format(Upper case)  

In [24]:
print("The intEger number is: {}".format(123))
print("The intEger number is: {:d}".format(123))
print("The intEger number is: {:f}".format(123))
print("The intEger number is: {:.2f}".format(123.234235))
print("The intEger number is: {:b}".format(123))
print("The intEger number is: {:o}".format(123))
print("The intEger number is: {:X}".format(123))
print("The intEger number is: {:x}".format(123))

The intEger number is: 123
The intEger number is: 123
The intEger number is: 123.000000
The intEger number is: 123.23
The intEger number is: 1111011
The intEger number is: 173
The intEger number is: 7B
The intEger number is: 7b


**Case 4 : Padding**  

In [40]:
print("The intEger number is: {:5d}".format(124563))
print("The intEger number is: {:9d}".format(124563))
print("The float number is: {:8f}".format(123.4567))
print("The float number is: {:08.3f}".format(123.4567))
print("The float number is: {:08.3f}".format(3.47))
print("The float number is: {:8.3f}".format(3.47))

The intEger number is: 124563
The intEger number is:    124563
The float number is: 123.456700
The float number is: 0123.457
The float number is: 0003.470
The float number is:    3.470


**{:08.3f}**
- Total positions should be minimum 8.
- After decimal point exactly 3 digits are allowed.If it is less then 0s will be placed in the last
positions
- If total number is < 8 positions then 0 will be placed in MSBs
- If total number is >8 positions then all integral digits will be considered.
- The extra digits we can take only 0  
  
**Case-4: Number formatting with alignment**  
<pre>
<      Left Alignment to the remaining space
^      Center alignment to the remaining space
>      Right alignment to the remaining space
=      Forces the signed(+) (-) to the left most position
</pre>

**Note:** Default Alignment for numbers is Right Alignment.

In [51]:
print("1.){:5d}".format(12))
print("2.){:<5d}".format(12))
print("3.){:<05d}".format(12))
print("4.){:>5d}".format(12))
print("5.){:>05d}".format(12))
print("6.){:^5d}".format(12))
print("7.){:=5d}".format(-12))
print("8.){:^10.3f}".format(12.23456))
print("9.){:=8.3f}".format(-12.23456))

1.)   12
2.)12   
3.)12000
4.)   12
5.)00012
6.) 12  
7.)-  12
8.)  12.235  
9.)- 12.235


**Case-5: String formatting with format()**  
Similar to numbers, we can format String values also with format() method  
**Note:** Default Alignment for string is left Alignment.

In [54]:
print("1){:5d}".format(12))
print("2){:5}".format("rat"))
print("3){:>5}".format("rat"))
print("4){:<5}".format("rat"))
print("5){:^5}".format("rat"))
print("6){:*^10}".format("rat")) #Instead of * we can use any character(like +,$,a etc) 

1)   12
2)rat  
3)  rat
4)rat  
5) rat 
6)***rat****


**Case-6: Truncating Strings with format() method**  

In [59]:
print("{:.4}".format("Ravinder Singh"))
print("{:5.4}".format("Ravinder Singh"))
print("{:>5.4}".format("Ravinder Singh"))
print("{:^5.4}".format("Ravinder Singh"))
print("{:*^5.4}".format("Ravinder Singh")) 

Ravi
Ravi 
 Ravi
Ravi 
Ravi*


**Case-7: Formatting dictionary members using format()**  

In [63]:
person={'age':28,
        'name':'Ravi'}

print("{p[name]}'s age is: {p[age]}".format(p=person))

Ravi's age is: 28


**Note: p is alias name of dictionary**  
<pre>More convinient way is to use **person</pre>

In [64]:
person={'age':28,
        'name':'Ravi'}

print("{name}'s age is: {age}".format(**person))

Ravi's age is: 28


**Case-8: Formatting class members using format()**      

In [66]:
class Person:
    age=28
    name="Ravi"
print("{p.name}'s age is :{p.age}".format(p=Person())) 

Ravi's age is :28


In [68]:
class Person:
    def __init__(self,name,age):
        self.name=name
        self.age=age
print("{p.name}'s age is :{p.age}".format(p=Person('Ravi',28)))
print("{p.name}'s age is :{p.age}".format(p=Person('Rohit',30))) 

Ravi's age is :28
Rohit's age is :30


**Case-9: Dynamic Formatting using format()**  

In [69]:
string="{:{fill}{align}{width}}"
print(string.format('RAVI',fill='*',align='^',width=5))
print(string.format('RAVI',fill='*',align='^',width=6))
print(string.format('RAVI',fill='*',align='<',width=6))
print(string.format('RAVI',fill='*',align='>',width=6)) 

RAVI*
*RAVI*
RAVI**
**RAVI


**Case-10: Dynamic Float format template**

In [70]:
num="{:{align}{width}.{precision}f}"
print(num.format(123.236,align='<',width=8,precision=2))
print(num.format(123.236,align='>',width=8,precision=2)) 

123.24  
  123.24


**Case-11: Formatting Date values**  

In [73]:
import datetime
#datetime formatting
date=datetime.datetime.now()
print(date)
print("It's now :{:%d/%m/%Y %H:%M:%S}".format(date)) 

2019-06-19 10:46:04.783929
It's now :19/06/2019 10:46:04


**Important Programs Regarding String Concept**  
**Q1. Write a program to reverse the given String**

In [30]:
#1st Way:
s=input("Enter Some String:")
print(s[::-1]) 

#2nd Way:
s=input("Enter Some String:")
print(''.join(reversed(s)))  #''.join(list(reversed(s)))

#3rd Way:
s=input("Enter Some String:")
i=len(s)-1
target=''
while i>=0:
    target=target+s[i]
    i=i-1
print(target)

Enter Some String:Learning Python is very easy!!
!!ysae yrev si nohtyP gninraeL
Enter Some String:Ravinder
rednivaR
Enter Some String:Ravinder Singh
hgniS rednivaR


**Q2 Program to reverse order of words.**

In [36]:
#input: Learning Python is very Easy
#output: Easy Very is Python Learning

s=input("Enter Some String:")
l=s.split()
l1=[]
i=len(l)-1
while i>=0:
    l1.append(l[i])
    i=i-1
output=' '.join(l1)
print(output) 

Enter Some String:Learning Python is very Easy
Easy very is Python Learning


**Q3. Program to reverse internal content of each word.**  

In [38]:
#input: Learning Python is very Easy
#output: gninraeL nohtyP si yrev ysaE

s=input("Enter Some String:")
l=s.split()
li=[]
i=0
while i<len(l):
    li.append(l[i][::-1])
    i=i+1
output=' '.join(li)
print(output) 

Enter Some String:gninraeL nohtyP si yrev ysaE
Learning Python is very Easy


**Q4. Write a program to sort the characters of the string and first alphabet symbols
followed by numeric values**  
input: B4A1D3  
Output: ABD134  


In [6]:
s=input("Enter Some String:")
s1=s2=output=''
for x in s:
    if x.isalpha():
        s1=s1+x
    else:
        s2=s2+x
for x in sorted(s1):
    output=output+x
for x in sorted(s2):
    output=output+x
print(output) 

Enter Some String:B4A1D3
ABD134


**Q5. Write a program for the following requirement**  
input: a4b3c2  
output: aaaabbbcc  

In [7]:
s=input("Enter Some String:")
output=''
for x in s:
    if x.isalpha():
        output=output+x
        previous=x 
    else:
        output=output+previous*(int(x)-1)
print(output) 

Enter Some String:a4b3c2
aaaabbbcc


**Q6. Write a program to perform the following activity**  
input: a4k3b2  
output:aeknbd  
  
**Note:**   
- chr(unicode)===>The corresponding character  
- ord(character)===>The corresponding unicode value  
```python
>>> ord('A')
>>> 65
>>> chr(65)
>>> 'A'
```

In [8]:
s=input("Enter Some String:")
output=''
for x in s:
    if x.isalpha():
        output=output+x
        previous=x
    else:
        output=output+chr(ord(previous)+int(x))
print(output)

Enter Some String:a4k3b2
aeknbd


**Q7 Write a program to remove duplicate characters from the given input string?**  
input: ABCDABBCDABBBCCCDDEEEF  
output: ABCDEF  

In [13]:
s=input("Enter Some String:")
l=[]
for x in s:
    if x not in l:
        l.append(x)
output=''.join(l)
print(output) 

Enter Some String:ABCDABBCDABBBCCCDDEEEF
ABCDEF


In [15]:
from collections import OrderedDict
foo = "mppmt"
"".join(OrderedDict.fromkeys(foo))

'mpt'

**Q8  Write a program to find the number of occurrences of each character present in the
given String?**   
input: ABCABCABBCDE  
output: A-3,B-4,C-3,D-1,E-1

In [16]:
s=input("Enter the Some String:")
d={}
for x in s:
    if x in d.keys():
        d[x]=d[x]+1
    else:
        d[x]=1 
for k,v in d.items():
    print("{} = {} Times".format(k,v)) 

Enter the Some String:ABCABCABBCDE
A = 3 Times
B = 4 Times
C = 3 Times
D = 1 Times
E = 1 Times
